enterprise/com/src/main/java/org/neo4j/com/storecopy/TransactionCommittingResponseUnpacker.java

```diff
@@ -191,7 +201,86 @@ public void unpackResponse( Response<?> response, final TxHandler txHandler ) th
+                applyQueuedTransactionsIfNeeded();
+            }
+        }
+    }
+
+    private void applyQueuedTransactionsIfNeeded() throws IOException
+    {
+        if ( transactionQueue.isEmpty() )
+        {
+            return;
+        }
+
+        /**
+         * Problem 1 (Not really a problem but we thought it where)
+         *  - chunk of batch is smaller than safe zone
+         *  - tx started after activeTransactions() is called
+         *  is safe because those transactions will see the latest state of store before chunk is applied and
+         *  because chunk is smaller than safe zone we are guarantied to not see two different states of any record
+         *  when applying the chunk.
+         *
+         *    activeTransactions() is called
+         *    |        start committing chunk
+         * ---|----+---|--|------> TIME
+         *         |      |
+         *         |      Start applying chunk
+         *         New tx starts here. Does not get terminated because not among active transactions, this is safe.
+         *
+         * Problem 2
+         *  - chunk of batch is larger than safe zone
+         *  - tx started after activeTransactions() but before apply
+         *
+         *    activeTransactions() is called
+         *    |        start committing chunk
+         * ---|--------|+-|------> TIME
+         *              | |
+         *              | Start applying chunk
+         *              New tx starts here. Does not get terminated because not among active transactions, but will read
+         *              outdated data and can be affected by reuse contamination.
+         */
+
+        // We stop new transactions from starting to avoid problem 1
+        if ( batchSizeExceedsSafeZone() )
+        {
+            // Problem 2
+            kernelTransactions.blockNewTransactions();
+        }
+        try
+        {
+            markUnsafeTransactionsForTermination();
+            applyQueuedTransactions();
+        }
+        finally
+        {
+            kernelTransactions.unblockNewTransactions();
+        }
+    }
+
+    private boolean batchSizeExceedsSafeZone()
+    {
+        long lastAppliedTimestamp = transactionQueue.last().getCommitEntry().getTimeWritten();
+        long firstAppliedTimestamp = transactionQueue.first().getCommitEntry().getTimeWritten();
+        long chunkLength = lastAppliedTimestamp - firstAppliedTimestamp;
+
+        return chunkLength > idReuseSafeZoneTime;
+    }
+
+    private void markUnsafeTransactionsForTermination()
+    {
+        long lastAppliedTimestamp = transactionQueue.last().getCommitEntry().getTimeWritten();
+        long earliestSafeTimestamp = lastAppliedTimestamp - idReuseSafeZoneTime;
+
+        for ( KernelTransaction tx : kernelTransactions.activeTransactions() )
+        {
+            KernelTransactionImplementation kti = (KernelTransactionImplementation) tx;
+            long commitTimestamp = kti.lastTransactionTimestampWhenStarted();
```

Comment:
Also, `commitTimestamp` and `lastTransactionTimestampWhenStarted` sounds like two different things.